### PR TITLE
Read Only Mode

### PR DESCRIPTION
**Pull request by quaxsze**

### CODE REVIEW

**Comment by abulte** [comment]
What we should block:
- new dataset
- new user (signup)
- new thread (discussion)
- new message (discussion)
- new reuse
- new community resource
- new organization
- new anomaly
- new follow (?)
- new harvester (?)

About feedback:
- we should display a banner both on front and admin when the mode is activated, explaining what's going on
- on admin side: a toast notification when an action fails because of the read only mode should be enough
- on front side: ideally we should hide or disable the buttons leading to blocked actions

Ideally, site admins should still be allowed to make all actions.

File: udata/settings.py [hunk 1]
```diff
@@ -446,6 +446,23 @@ class Defaults(object):
 
     API_DOC_EXTERNAL_LINK = 'https://doc.data.gouv.fr/api/reference/'
 
+    # Read Only Mode 
+    ####################
+    # This mode can be used to stop a spam attack.
```
Comment (abulte) [inline udata/settings.py:451]:
```suggestion
    # This mode can be used to mitigate a spam attack for example.
```

**Review by abulte — COMMENTED** [review]
Nice!

We should document the use of the `423` code: why we chose this one. We also should update the API doc on doc.data.gouv.fr.

**Review by abulte — COMMENTED** [review]
Maybe we should call this `PROTECTION_MODE`?

File: udata/api/__init__.py [hunk 2]
```diff
@@ -100,6 +100,9 @@ def _apply_secure(self, func, permission=None):
 
         @wraps(func)
         def wrapper(*args, **kwargs):
+            if current_app.config['READ_ONLY_MODE'] and any(ext in str(func) for ext in current_app.config['METHOD_BLOCKLIST']):
```
Comment (abulte) [inline udata/api/__init__.py:103]: Can we add a super admin condition here? Super admins should always be allowed to do everything.

**Review by abulte — COMMENTED** [review]
We should add a banner on the admin side too.

File: js/config.js [hunk 3]
```diff
@@ -165,6 +165,11 @@ export const is_search_autocomplete_enabled = _jsonMeta('search-autocomplete-ena
  */
 export const search_autocomplete_debounce = _jsonMeta('search-autocomplete-debounce');
 
+/**
+ * Whether the 'delete me' feature is enabled or not.
```
Comment (abulte) [inline js/config.js:169]: fixme :-)

**Comment by restuccia** [comment]
The contribution view could be hidden (it currently works and redirects users to the admin page, where the small banner on the bottom right corner is shown)

Capture d’écran 2020-11-23 à 10 38 59

**Comment by abulte** [comment]
@quaxsze a few corner cases to handle for unprivileged and then it will be good I think 😎 

- new anomaly: 423 OK _mais_ le bouton est toujours là et l'erreur n'est pas gérée : enlever le bouton front
- new follow: 423 OK _mais_ le bouton est toujours là et l'erreur n'est pas gérée : enlever le bouton front
- contacter le producteur : 423 OK _mais_ le bouton est toujours là et l'erreur n'est pas gérée : enlever le bouton front

and maybe this one if it's not too much of a PITA (we can do w/o):

- new harvester: 423 mais pas le bon message d'erreur https://sentry.data.gouv.fr/etalab/demo-datagouvfr/issues/1873131/events/9854130/

File: js/admin.vue [hunk 4]
```diff
@@ -44,7 +45,16 @@ export default {
...
     },
     ready() {
-        // Display an error identifier un uncaught error
+        // Displays a toast if read only mode is enabled
+        if (this.readOnlyEnabled) {
+            this.notifications.push({
+                type: 'error',
+                icon: 'exclamation-triangle',
+                title: this._('Attention'),
+                details: this._('Due to unusual activities, the creation of new content is currently disabled.'),
```
Comment (abulte) [inline js/admin.vue:54]:
```suggestion
                details: this._('Due to security reasons, the creation of new content is currently disabled.'),
```

Let's revert this, cf Mario's comment, my bad.

File: js/admin.vue [hunk 5]
```diff
@@ -44,7 +45,16 @@ export default {
...
+            });
+        }
+        // Displays an error identifier un uncaught error
```
Comment (abulte) [inline js/admin.vue:57]:
```suggestion
        // Displays an error identifier on uncaught error
```

File: udata/api/__init__.py [hunk 6]
```diff
@@ -100,6 +100,12 @@ def _apply_secure(self, func, permission=None):
 
         @wraps(func)
         def wrapper(*args, **kwargs):
+            if not current_user.is_anonymous and \
+                not current_user.sysadmin and \
+                current_app.config['READ_ONLY_MODE'] and \
+                any(ext in str(func) for ext in current_app.config['METHOD_BLOCKLIST']):
```
Comment (abulte) [inline udata/api/__init__.py:106]:
```suggestion
            if (
                not current_user.is_anonymous and
                not current_user.sysadmin and
                current_app.config['READ_ONLY_MODE'] and 
                any(ext in str(func) for ext in current_app.config['METHOD_BLOCKLIST'])
            ):
```

Personal preference, YMMV

File: udata/api/__init__.py [hunk 7]
```diff
@@ -100,6 +100,12 @@ def _apply_secure(self, func, permission=None):
...
+                self.abort(423, 'Due to unusual activities, the creation of new content is currently disabled.')
```
Comment (abulte) [inline udata/api/__init__.py:107]:
```suggestion
                self.abort(423, 'Due to security reasons, the creation of new content is currently disabled.')
```

File: udata/auth/forms.py [hunk 8]
```diff
@@ -14,6 +15,12 @@ class ExtendedRegisterForm(RegisterForm):
     last_name = fields.StringField(
         _('Last name'), [validators.DataRequired(_('Last name is required')),
                          validators.NoURLs(_('URLs not allowed in this field'))])
+    
+    def validate(self):
+        if not super().validate() or current_app.config.get('READ_ONLY_MODE'):
```
Comment (abulte) [inline udata/auth/forms.py:20]:
```suggestion
        # no register allowed when read only mode is on
        if not super().validate() or current_app.config.get('READ_ONLY_MODE'):
```

**Review by abulte — APPROVED** [review]
Missing CHANGELOG entry and doc in `adapting-settings.md`

**Review by abulte — COMMENTED** [review]
Also, extract i18n strings before merging

**Pull request merged**